### PR TITLE
[Keyboard] add Keychron K4V2 RGB

### PR DESCRIPTION
**Pull request by coffeec0der**
# Description

added k4

# Checklist

- [x] functioning keyboard
- [ ] full original functionality (Bluetooth not working, battery check shortcut not working, no macos support)
- [x] RGB
- [x] VIA

### CODE REVIEW

File: keyboards/keychron/k4/rgb/v2/config.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+/* LED Status indicators */
+
+#include <config_led.h>
+
+
+#define DIP_SWITCH_PINS { D5, D6 }
+#define LED_CAPS_LOCK_PIN B12
+#define LED_NUM_LOCK_PIN  D4
+#define LED_PIN_ON_STATE  1
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/config.h]: those need moving in data driven format in `info.json`

```c
    "indicators": {
        "caps_lock": "B12",
        "num_lock": "D4",
        "on_state": 1
    },
```

File: keyboards/keychron/k4/rgb/v2/config.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#define RGB_MATRIX_LED_COUNT 100
+
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 19
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/config.h]: not needed anymore
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/rules.mk [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# project specific files
+#SRC = ../../drivers/led/sn32/matrix_sn32f24xx.c
+#SRC += config_led.c
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/rules.mk]: stale, safe to remove
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/via/config.h [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
+#pragma once
+
+#define WEAR_LEVELING_SN32_EMULATION_PAGE_COUNT 32
+#define LAYER_STATE_8BIT
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/via/config.h]: there shouldn't be a reason to define this
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/via/config.h [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+#define DYNAMIC_KEYMAP_LAYER_COUNT 2
+#define DYNAMIC_KEYMAP_EEPROM_MAX_ADDR 1023
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/via/config.h]: with the wear leveling eeprom driver, there should be no reason to mess with addresses
also, 4 dynamic layers are probably possible, by bumping `WEAR_LEVELING_SN32_EMULATION_PAGE_COUNT`

File: keyboards/keychron/k4/rgb/v2/readme.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+* Hardware Supported: *Keychron K4V2 RGB, MCU: SN32F248BF*
+* Hardware Availability: *https://www.keychron.com/products/keychron-k4-wireless-mechanical-keyboard-version-2*
+
+Make example for this keyboard (after setting up your build environment):
+
+    qmk compile -kb keychron/k4/rgb/v2/ -km ansi
+
+Make example for this keyboard (after setting up your build environment):
+
+    qmk compile -kb keychron/k4/rgb/v2/via -km ansi
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/readme.md]: this kind of structure is a bit oldschool and dirty. The proper way to do it is to have via support in a via keymap.
with KC boards being both ANSI and ISO, it makes sense to treat them as variants. eg
`keychron/k4/{rgb,white}/{v1,v2}/{ansi,iso}/keymaps/{default,via}/keymap.c`

File: keyboards/keychron/k4/rgb/v2/via/readme.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Keychron K4v2
+
+K4v2 Gateron RGB with VIA support
+
+Make example for this keyboard (after setting up your build environment):
+
+    make keychron/k4/v2/rgb/via:default
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/via/readme.md]: should be updated to `qmk  compile -kb`

**Review by dexter93 — COMMENTED** [review]
first pass on code review. nice work so far!

**Comment by dexter93** [comment]
please satisfy the linter
```
linting keychron/k4/rgb/v2/ansi
☒ keychron/k4/rgb/v2/ansi: The file "keyboards/keychron/k4/rgb/v2/ansi/led.c" does not have a license header!
☒ keychron/k4/rgb/v2/ansi/via: The file "keyboards/keychron/k4/rgb/v2/ansi/keymaps/via/config.h" does not have a license header!
☒ Lint check failed for: keychron/k4/rgb/v2/ansi
File 'keyboards/keychron/k4/rgb/v2/ansi/keymaps/via/via_ansi.json' Requires Formatting
```

File: keyboards/keychron/k4/rgb/v2/ansi/config.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+
+/* LED Status indicators */
+
+
+#define DIP_SWITCH_PINS { D5, D6 }
+#define RGB_MATRIX_LED_COUNT 100
+
+
+#define BACKLIGHT_LEVELS 8
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/config.h]: that doesn't do anything atm.
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/ansi/config.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 19
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/config.h]: these get autocalculated from the info.json, no need to specify them
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/ansi/config.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+#define LED_MATRIX_ROWS MATRIX_ROWS
+#define LED_MATRIX_ROW_CHANNELS 3
+#define LED_MATRIX_ROWS_HW (LED_MATRIX_ROWS * LED_MATRIX_ROW_CHANNELS)
+#define LED_MATRIX_ROW_PINS { C3, C1, C0, C6, C5, C4, C9, C8, C7, C12, C11, C10, B13, C14, C13, B14, B15, D3 }
+#define MATRIX_COL_PINS { A8, A9, A10, A11, A12, A13, A14, A15, B0, B1, B2, B3, B4, B5, B6, B7, B8, B9, B10 }
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/config.h]: this one matches the
```json
    "matrix_pins": {
        "cols": ["A8", "A9", "A10", "A11", "A12", "A13", "A14", "A15", "B0", "B1", "B2", "B3", "B4", "B5", "B6", "B7", "B8", "B9", "B10"],
```
so it better be in one place
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/ansi/config.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+
+#define BACKLIGHT_LEVELS 8
+
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 19
+
+#define LED_MATRIX_ROWS MATRIX_ROWS
+#define LED_MATRIX_ROW_CHANNELS 3
+#define LED_MATRIX_ROWS_HW (LED_MATRIX_ROWS * LED_MATRIX_ROW_CHANNELS)
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/config.h]: the driver will include those now as defaults, safe to omit
```suggestion
```

File: keyboards/keychron/k4/rgb/v2/ansi/config.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+#define LED_MATRIX_ROW_PINS { C3, C1, C0, C6, C5, C4, C9, C8, C7, C12, C11, C10, B13, C14, C13, B14, B15, D3 }
+#define MATRIX_COL_PINS { A8, A9, A10, A11, A12, A13, A14, A15, B0, B1, B2, B3, B4, B5, B6, B7, B8, B9, B10 }
+
+#define LED_MATRIX_COLS MATRIX_COLS
+#define LED_MATRIX_COL_PINS MATRIX_COL_PINS
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/config.h]: the driver will include those now as defaults, safe to omit

```suggestion
```

File: keyboards/keychron/k4/rgb/v2/ansi/led.c [hunk 13]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2023 PythonDeployer (@PythonDeployer)
```
Comment (dexter93) [inline keyboards/keychron/k4/rgb/v2/ansi/led.c]: this file does not get included at build time. Why do you need it?

**Review by dexter93 — CHANGES_REQUESTED** [review]

File: keyboards/keychron/k4/rgb/v2/ansi/led.c (hunk 13 above)
Comment (coffeec0der) [reply keyboards/keychron/k4/rgb/v2/ansi/led.c]: Idk i guess added it with debugging and forgot to remove, ill delete cuz not needed :)

**Pull request merged**